Staleness

Pokom opened the pull request:
#6 

I believe this can be used as a basis for staleness for a user. The effective flow is when the init is called, we set up an event listener for the document.hidden function. When this event is emitted, we pass in a function that will check to see if the user is hidden or not, and emit an appropriate event to the Clique. The user isn't immediately set to being stale, they first emit an event stating they've gone hidden, and then after 30 seconds of being hidden they emit a message saying they've gone stale. Once this happens the FriendsList updates the icon to be a green icon instead of the default.

My proposed features to enhance this since it's still a little immature, is to animate a countdown on the friends icons to show how much longer before the user goes stale. 

I believe the only code that's messy is in init.js. I am open to ways of improving that code so it's more modularized.

Pokom committed:
Initial commit for the staleness

I'm able to set a flag when a person goes stale, but I'd ideally like
to notifiy all of the users this person stepped away and then
begin a countdown to mark them as being inactive. It'd be cool to change their marker
and have it countdown to show how much longer until the user times out.

Pokom committed:
Got the hidden to start working properly.

Time to move on and have it timed. What I think I'm going to do is have it delay the broadcast
for 30 seconds or so. Might warn users someone is going idle, but ultimately have it hold
off on going stale for a certain amount of time.

Pokom committed:
Finishing up the updates when a user goes into hiding

Added event listener to communicate when a user starts going hidden and then
after 30 seconds emits an event saying they've gone stale. Still have some more work
to do to clean up how it looks when they go stale. Also probably would love to have an
animation that changes the icon from red to gray over 30 seconds to showcase the
user going inactive.

Pokom committed:
Split up functionality for updating the marker

Pokom committed:
Removing a trailing,

Pokom commented on client/Friend.js:
```diff
@@ -68,4 +68,9 @@ Friend.prototype.update = function(info) {
+	this.updateMarker(info.stale)
+}
+Friend.prototype.updateMarker = function(stale) {
+	var icon = stale ? 'http://maps.google.com/mapfiles/ms/icons/green.png' : ''
```
I would like to get a better icon to use for when it's stale. It looks pretty bad.

ChrisMondok replied:
I've got a branch that generates icons on the client, so this is easily resolved.

Pokom replied:
Then I'll remove this and defer responsibility to the client

Pokom commented on client/FriendsList.js:
```diff
@@ -16,12 +16,24 @@ FriendsList.prototype.update = function(message) {
+			break
+		case 'friend-stale':
+			new Toast(message.name+' Is no longer active')
+			this.handleUpdate(message)
```
We can just illicit an update here because the message will have the stale flag being set.

Pokom commented on client/init.js:
```diff
@@ -4,9 +4,9 @@ addEventListener('load', function() {
 	var destinationView = null
-
+	var timeoutId = null
+	var timeout = 30000 //30s
```
I think we should tweek this a bit  to make it a bit more user friendly. 30 seconds works well, but I don't know if that is too much/little time.

Pokom commented on client/init.js:
```diff
@@ -87,17 +88,45 @@ addEventListener('load', function() {
 		})
 	}
 
+	function setupHiddenEvent(){
```
This entire function feels dirty to me, but it does what it needs to. Do you have any thoughts on how to improve this?

ChrisMondok replied:
I don't know how hard this would be, but what if the Visibility object took the comms object as an argument, and did all of its event firing in there? That is, get it all out of init.js

Pokom replied:
Yeah I'm down with that. Date night so I'll work on it tomorrow.

Pokom commented on client/visibility.js:
```diff
@@ -0,0 +1,23 @@
+//heavy influence: https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API
```
This visibility object is useful for cross browser functionality. Who would've thought each browser has their own specific hidden flag.

ChrisMondok replied:
Inconsistent APIs? Who'd have thought‽

Pokom committed:
Removed the entire conditional for going stale

Pokom committed:
Removed unused event in Clique

Pokom committed:
Fixed a minor defect introduced with the staleness.

When a new user joined, the senders state wasn't saved accordingly. This caused a new user
to not see the old users position/staleness until an update was sent

Pokom commented on Cliques.js:
```diff
@@ -62,9 +59,12 @@ Clique.prototype.onMessage = function(sender, message) {
 		switch (payload.event) {
 			case 'friend-updated':
-				sender.position = payload.position
+				sender.position = !payload.position ? sender.position : payload.position
```
This line is unfortunately necessary as of now because the only time the position is broadcasted is on updates. When the user goes stale, we do not have access to their previous position(at least I don't think so).

ChrisMondok commented on client/init.js:
```diff
@@ -87,17 +88,45 @@ addEventListener('load', function() {
+			if (hiding){
+				timeoutId = window.setTimeout(function(){
+					sendMessage({
+						event: 'friend-updated',
```
What's the timeout for? Also, I don't like the client being authoritative as to whether or not it's stale. That seems like a display concern.

Pokom replied:
So my thinking was to have two separate events. One is when they go hidden, notify users they're hidden. After the timeout is over itll broadcast that the users location is stale. I can figure out how to get it into the appropriate display

ChrisMondok commented on client/init.js:
```diff
@@ -87,17 +88,45 @@ addEventListener('load', function() {
+			else
+				window.clearTimeout(timeoutId)
+			sendMessage({
+				event: 'friend-updated',
```
I'd give this a different event. Something like friend-activity-state-changed or something less unwieldy. This way, everyone doesn't need to check whether properties are found or not on the message.

Basically, each message should act like a well known type, where if you get a message with the type you care about, it's properties should be accessible. Maybe we should make types for them and share them somehow...

Pokom replied:
I like how redux handles event propagation, but it's heavily reliant on module imports on both client and server.

ChrisMondok commented on client/FriendsList.js:
```diff
@@ -42,6 +42,7 @@ FriendsList.prototype.handleUpdate = function(message) {
 	var friend = this.friends.find({id: message.id})
 	if(friend) friend.update(message)
 	else this.addFriend(new Friend(this.map, message))
+	if(message.stale) new Toast(message.name + ' has gone stale')
```
These toasts are going to be pretty frequent if a user is going back and forth from this app and another. We should probably just change the icon and be done with it.

Pokom replied:
Yeah I agree. Ill remove it.

Pokom committed:
Taking comments from PR and updating where the communication is sent

When initializing the visibility object, I pass in the communications object.

I removed the timeout for simplicity for now, once the user's session goes into
the background it will emit an event stating the user went state.

Removed the toast for when a user goes hidden.

Removed the updateMarker function and will defer the icon to the client.
